pumpp/task/key.py

```diff
@@ -0,0 +1,438 @@
+#!/usr/bin/env python
+# -*- encoding: utf-8 -*-
+'''Key recognition task transformer'''
+
+from itertools import product
+
+import numpy as np
+import mir_eval
+import jams
+
+from librosa import note_to_midi, midi_to_note, time_to_frames
+from librosa.sequence import transition_loop
+
+from .base import BaseTaskTransformer
+from ..exceptions import ParameterError
+from ..labels import LabelBinarizer, LabelEncoder, MultiLabelBinarizer
+
+__all__ = ['KeyTransformer', 'KeyTagTransformer']
+
+C_MAJOR = ['C', 'D', 'E', 'F', 'G', 'A', 'B']
+C_MAJOR_PITCHES = note_to_midi(C_MAJOR) % 12
+MODES = ['ionian', 'dorian', 'phrygian', 'lydian', 'mixolydian', 'aeolian', 'locrian']
+QUALITY = {'major' : 0, 'minor' : -3}
+
+class KeyTransformer(BaseTaskTransformer):
+    '''Key annotation transformer.
+
+    This transformer uses a (pitch_profile, tonic) decomposition of key_mode 
+    annotation, where the mode is reflected in the 12-D pitch_profile vector.
+
+    Attributes
+    ----------
+    name : str
+        The name of the key trnsformer
+
+    sr : number > 0
+        The sampling rate of audio
+
+    hop_length : int > 0
+        The number of samples between each annotation frame
+
+    sparse : bool
+        If True, tonic value is sparsely encoded as integers in [0, 12].
+        If False, tonic value is densely encoded as 13-dimensional booleans.
+    '''
+    def __init__(self, name='key', sr=22050, hop_length=512, sparse=False):
+        '''Initialize a key task transformer'''
+
+        super(KeyTransformer, self).__init__(name=name,
+                                             namespace='key_mode',
+                                             sr=sr, hop_length=hop_length)
+        self.sparse = sparse
+ 
+        self.register('pitch_profile', [None, 12], np.bool)
+        if self.sparse:
+            self.register('tonic', [None, 1], np.int)
+        else:
+            self.register('tonic', [None, 13], np.bool)
+
+    def _encode_key_str(self, key_str):
```

Comment:
if this is a helper, i don't think it needs to be in the class.  It only accesses self for sparse; maybe you can make sparse a parameter, and pull the method out of the class?

Reply:
Sounds good. Will do.